### PR TITLE
added typescript to project

### PR DESCRIPTION
**Pull request by 96mickey**

### CODE REVIEW

File: app.ts [hunk 1]
```diff
@@ -0,0 +1,341 @@
+import { data } from "./data";
+import { validate } from "./validate";
+
+let button: HTMLElement = document.getElementById("changeColor");
+let table: HTMLElement = document.getElementById("tablebody");
+
+interface Usermdl {
```
Comment (samarpan-b) [inline app.ts]: Why this interface ?

File: app.ts [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+  "Middle Name": string;
+  "Last Name": string;
+  Email: string;
+  "Phone Number": number;
+  Address: string;
+  Role?: string;
+}
+
+//use this for Object.entries
+interface ObjectInterface {
```
Comment (samarpan-b) [inline app.ts]: No need of this. Object.entries will anyways be available.

File: app.ts [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+class User implements Usermdl {
+  "First Name": string;
+  "Middle Name": string;
+  "Last Name": string;
+  "Email": string;
+  "Phone Number": number;
+  "Address": string;
+  "Role": string;
+
+  constructor(
```
Comment (samarpan-b) [inline app.ts]: use partial here

File: app.ts [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+    this["First Name"] = firstname;
+    this["Middle Name"] = middlename;
+    this["Last Name"] = lastname;
+    this["Email"] = email;
+    this["Phone Number"] = phonenumber;
+    this["Address"] = address;
+  }
+}
+
+var loadData = (): void => {
```
Comment (samarpan-b) [inline app.ts]: no var please

File: app.ts [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+  address: HTMLInputElement;
+  number: HTMLInputElement;
+  email: HTMLInputElement;
+}
+
+// export interface Listener<T> {
+//   (event: T): any;
+// }
+
+class User implements Usermdl {
```
Comment (samarpan-b) [inline app.ts]: every class need to be in its own file.

File: app.ts [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+    this["Last Name"] = lastname;
+    this["Email"] = email;
+    this["Phone Number"] = phonenumber;
+    this["Address"] = address;
+  }
+}
+
+var loadData = (): void => {
+  //setting up the basic view
+  let items: object[] = [...data];
```
Comment (samarpan-b) [inline app.ts]: use actual user Interface or class here

File: app.ts [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+
+  //finally getting the edit button back
+  document.getElementById(
+    `edit_wrapper_${index}`
+  ).innerHTML = `<button id="edit_${index}" class="btn btn-secondary" >edit</button>`;
+  document.getElementById(`edit_${index}`).addEventListener("click", editRow);
+  // }
+};
+
+var makeLayout = (items: object[]): string[] => {
```
Comment (samarpan-b) [inline app.ts]: why object ?

File: app.ts [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+  <td colspan=2><button class="btn btn-primary btn-block" type="button" id="submitnewuser" >Add</button></td>
+  </tr>`;
+
+  itemsToDisplay.push(inputelement);
+
+  return itemsToDisplay;
+};
+
+//general way for selecting elements to iterate over
+export var callElements = (
```
Comment (samarpan-b) [inline app.ts]: why export here ?

File: data.ts [hunk 9]
```diff
@@ -0,0 +1,26 @@
+export var data: object[] = [
```
Comment (samarpan-b) [inline data.ts]: no var pls

**Review by samarpan-b — CHANGES_REQUESTED** [review]
where is webpack config ?

File: app.ts (hunk 8 above)
Comment (96mickey) [reply app.ts]: I am using this in validate.ts

> where is webpack config ?



> where is webpack config ?

Resolved. That file was skipped as it was a *.js file

**Comment by 96mickey** [comment]
> where is webpack config ?

Resolved. That file was skipped as it was a *.js file

File: app.ts (hunk 7 above)
Comment (96mickey) [reply app.ts]: Because I am passing an array of objects as it's first parameter.

File: app.ts (hunk 8 above)
Comment (96mickey) [reply app.ts]: Actually I was planning to use this in validate module, but then i skipped that part.

File: app.ts [hunk 10]
```diff
@@ -0,0 +1,309 @@
...
+}
+
+// type ListUser = [UserItem];
+
+interface UserItems extends Array<UserItem> {}
+
+let loadData = (): void => {
+  //setting up the basic view
+  // let items = Array as {} as UserItem
+  let items = <UserItems>[...data];
```
Comment (samarpan-b) [inline app.ts]: fix this

File: app.ts [hunk 11]
```diff
@@ -0,0 +1,309 @@
+import { data } from "./data";
+import { validate } from "./validate";
+import { User } from "./user";
+
+let button: HTMLElement = document.getElementById("changeColor");
+let table: HTMLElement = document.getElementById("tablebody");
+
+interface UserHtmlElements {
```
Comment (samarpan-b) [inline app.ts]: move these to separate file

File: app.ts [hunk 12]
```diff
@@ -0,0 +1,309 @@
...
+interface UserItems extends Array<UserItem> {}
+
+let loadData = (): void => {
+  //setting up the basic view
+  // let items = Array as {} as UserItem
+  let items = <UserItems>[...data];
+
+  let itemsToDisplay: string[] = makeLayout(items);
+
+  table.innerHTML = itemsToDisplay.join("");
```
Comment (samarpan-b) [inline app.ts]: try to use OOP concepts

File: app.ts [hunk 13]
```diff
@@ -0,0 +1,309 @@
...
+  let index: object = /\d+/.exec(e.path[0].id);
+  let elements: object = callElements(index[0], "");
+
+  let editbtn = document.getElementById(`edit_wrapper_${index[0]}`);
+
+  //hiding the elements of the table
+  //and changing the row to input fields. setting up the initial value
+  //and adding event to validate the input fields
+  let regex = /\_(.*?)\_/;
+  (<any>Object).entries(elements).forEach(<T>(entry: T) => {
```
Comment (samarpan-b) [inline app.ts]: remove any

File: app.ts [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+  }">Save</button><button id="cancel_${
+    index[0]
+  }" class="btn btn-outline-warning" >Cancel</button>`;
+  let savebtn: HTMLElement = document.getElementById(`save_${index[0]}`);
+  let cancelbtn: HTMLElement = document.getElementById(`cancel_${index[0]}`);
+  savebtn.addEventListener("click", save);
+  cancelbtn.addEventListener("click", cancel);
+};
+
+const cancel = (e: any): void => {
```
Comment (samarpan-b) [inline app.ts]: use browser event

**Review by samarpan-b — CHANGES_REQUESTED** [review]

File: types.ts [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+  fname: HTMLInputElement;
+  mname: HTMLInputElement;
+  lname: HTMLInputElement;
+  address: HTMLInputElement;
+  number: HTMLInputElement;
+  email: HTMLInputElement;
+}
+
+export interface UserItem {
+  "First Name": string;
```
Comment (samarpan-b) [inline types.ts]: no space please

File: data.ts [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+  {
+    "First Name": "Mayank",
+    "Middle Name": " ",
+    "Last Name": "Nauriyal",
+    Email: "test@email.com",
+    "Phone Number": "8888888888",
+    Address: "ABC street, XYZ place"
+  },
+  {
+    "First Name": "Mickey",
```
Comment (samarpan-b) [inline data.ts]: no space please. Only camelcase

**Review by samarpan-b — CHANGES_REQUESTED** [review]

File: userlist.ts [hunk 17]
```diff
@@ -0,0 +1,303 @@
...
+
+export class UserList {
+  "users": UserItem[] = [];
+
+  constructor(init?: UserItem[]) {
+    Object.assign(this.users, init);
+  }
+
+  loadData = () => {
+    let items: UserItem[] = [...this.users] as UserItem[];
```
Comment (samarpan-b) [inline userlist.ts]: no need

File: userlist.ts [hunk 18]
```diff
@@ -0,0 +1,303 @@
...
+
+    let itemsToDisplay: string[] = this.makeLayout(items);
+
+    table.innerHTML = itemsToDisplay.join("");
+
+    button.innerText = "Refresh data";
+    button.onclick = this.loadData;
+
+    //adding event to add new user
+    let submitnewuserbtn: HTMLElement = document.getElementById(
```
Comment (samarpan-b) [inline userlist.ts]: htmlbuttonelement

File: userlist.ts [hunk 19]
```diff
@@ -0,0 +1,303 @@
...
+      "submitnewuser"
+    );
+    submitnewuserbtn.addEventListener("click", this.addUser);
+
+    //adding event to validate user input
+    let elements: object = callElements("new", "input_");
```
Comment (samarpan-b) [inline userlist.ts]: no need of object

File: validate.ts [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+//     let error = false;
+
+//     if(!validate(index, "email") || !validate(index, "address") || !validate(index, "fname") || !validate(index, "lname") || !validate(index, "number"))
+//     error = true;
+
+//     return error;
+
+// }
+
+export const validate = (e: any): boolean => {
```
Comment (samarpan-b) [inline validate.ts]: no any please

**Review by samarpan-b — CHANGES_REQUESTED** [review]

File: userlist.ts (hunk 18 above)
Comment (96mickey) [reply userlist.ts]: Working on this!
Changing it to HTMLButtonElement is giving some error.

File: userlist.ts (hunk 19 above)
Comment (96mickey) [reply userlist.ts]: I am using this for input as well as normal div elements. div elements does not have value attribute, so using an interface here is giving issue.

File: validate.ts (hunk 20 above)
Comment (96mickey) [reply validate.ts]: This is an InputEvent, but when this event is not recognized.

**Review by samarpan-b — APPROVED** [review]

**Pull request merged**